[Perf][CSSimplify] Transfer conformance requirements of a parameter to an argument

xedin opened the pull request:
<!-- What's in this pull request? -->
Performance optimization to help rule out generic overload choices sooner.

This is based on an observation of the constraint solver behavior,
consider following example:

```swift
func test<U: ExpressibleByStringLiteral>(_: U) {}

test(42)
```

Constraint System for call to `test` is going to look like this:

```
$T_test := ($T_U) -> $T_U
$T_U <conforms to> ExpressibleByStringLiteral
$T_42 <argument conversion> $T_U
$T_42 <literal conforms to> ExpressibleByIntegerLiteral
```

Currently to find out that `test` doesn't match, solver would have
to first find a binding for `$T_42`, then find a binding for `$T_U`
(through `<argument conversion>` constraint) and only later fail
while checking conformance of `Int` to `ExpressibleByStringLiteral`.

Instead of waiting for parameter to be bound, let's transfer conformance
requirements through a conversion constraint directly to an argument type,
since it has to conform to the same set of protocols as parameter to
be convertible (restrictions apply i.e. protocol composition types).

With that change it's possible to verify conformances early and reject
`test<U: ExpressibleByStringLiteral>` overload as soon as type of an
argument has been determined. This especially powerful for chained calls
because solver would only have to check as deep as first invalid argument
binding.


<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin commented:
@swift-ci please smoke test compiler performance

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test compiler performance

xedin commented:
~70% scope reduction is something I really like...

LucianoPAlmeida commented:
> ~70% scope reduction is something I really like...

Wow, Awesome! 🚀

xedin commented:
Although this was exciting, it would break cases like this https://github.com/apple/swift/pull/36691 where implicit conversion is involved (as pointed out by @hborla).

xedin commented:
I'm not giving up on this one, I think that it should be possible to just check implicit conversions, although it means a bit more conformance checking it would still result in significant filtering advantage.

xedin committed:
[ConstraintSystem] Add a new transitive conformance constraint

Conformance constraints could be transferred through conversions,
but that would also require checking implicit conversions
such as optional and pointer promotions for conformance is the
type itself doesn't conform, for that let's add a special constraint
`TransitivelyConformsTo`.

xedin committed:
[CSSimplify] Transfer conformance requirements of a parameter to an argument

Performance optimization to help rule out generic overload choices sooner.

This is based on an observation of the constraint solver behavior,
consider following example:

```swift
func test<U: ExpressibleByStringLiteral>(_: U) {}

test(42)
```

Constraint System for call to `test` is going to look like this:

```
$T_test := ($T_U) -> $T_U
$T_U <conforms to> ExpressibleByStringLiteral
$T_42 <argument conversion> $T_U
$T_42 <literal conforms to> ExpressibleByIntegerLiteral
```

Currently to find out that `test` doesn't match, solver would have
to first find a binding for `$T_42`, then find a binding for `$T_U`
(through `<argument conversion>` constraint) and only later fail
while checking conformance of `Int` to `ExpressibleByStringLiteral`.

Instead of waiting for parameter to be bound, let's transfer conformance
requirements through a conversion constraint directly to an argument type,
since it has to conform to the same set of protocols as parameter to
be convertible (restrictions apply i.e. protocol composition types).

With that change it's possible to verify conformances early and reject
`test<U: ExpressibleByStringLiteral>` overload as soon as type of an
argument has been determined. This especially powerful for chained calls
because solver would only have to check as deep as first invald argument
binding.

xedin commented:
@swift-ci please smoke test compiler performance

xedin commented:
@swift-ci please test source compatibility

DougGregor commented on validation-test/Sema/type_checker_perf/fast/rdar46713933_concrete_args.swift:
```diff
@@ -7,8 +7,5 @@ func wrap<T: ExpressibleByFloatLiteral>(_ key: String, _ value: T) -> T { return
 func wrapped(i: Int) -> Int {
-  // FIXME: When this stops being "too complex", turn the integer value into
-  // an integer literal.
-  // expected-error@+1{{reasonable time}}
```
Heh, should we try doing what the FIXME suggests as well?

xedin replied:
Yeah, I have it as - https://github.com/apple/swift/blob/main/validation-test/Sema/type_checker_perf/slow/rdar46713933_literal_arg.swift. Still slow unfortunately since it can't use `Int` early (yet).

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test compiler performance

xedin committed:
[CSSimplify] NFC: Move protocol composition check to `simplifyTransitivelyConformsTo`

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test compiler performance

hamishknight commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -6186,6 +6218,91 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyConformsToConstraint(
+
+  auto *protocol = protocolTy->castTo<ProtocolType>()->getDecl();
+
+  // First, let's check whether the type itself conforms,
+  // if it does - we are done.
+  if (TypeChecker::conformsToProtocol(resolvedTy, protocol, DC))
+    return SolutionKind::Solved;
```
I think you may want to avoid checking conditional conformance requirements by using `lookupConformance` here and below, as it's possible an implicit conversion changes e.g the element type in an array-to-array or optional-to-optional conversion.

xedin replied:
I don't think it should for anything besides `String` maybe but yeah, it's might be better to under-check here and let the parameter handle the rest...

hamishknight replied:
I mean something like:

```
protocol P {}
extension Optional : P where Wrapped == Any {}

func bar<T : P>(_ x: T) -> T { x }
var i: Int?
let _: Any? = bar(i)
```

xedin replied:
Yeah, I understand.

hborla replied:
Just to clarify because the documentation of `lookupConformance` isn't super clear - will this return a valid `ProtocolConformanceRef` when checking if `Int?` conforms to `P` in the above example?

xedin replied:
Yes, `lookupConformance` would return a conformance for `Optional : P` with conditional requirements list.

hamishknight commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -6186,6 +6218,91 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyConformsToConstraint(
+          BoundGenericType::get(ptrDecl, /*parent=*/Type(), {resolvedTy}));
+      typesToCheck.push_back(BoundGenericType::get(
+          ctx.getUnsafeMutablePointerDecl(), /*parent=*/Type(), {resolvedTy}));
+    }
```
I think you'll also want to handle `AnyHashable` here

xedin replied:
Good catch! I actually have to add a test for that just like I did for `Optional` and `UnsafePointer`

xedin replied:
I'm going to let full suite run before pushing that though.

xedin replied:
Actually I wonder whether I do really need to check for `AnyHashable` here e.g.

```
enum E {
  case test(_: String)
}

func overloaded() -> Int { 42 }
func overloaded() -> E { E.test("") }

func test<T: Hashable>(_: T) -> T {
  fatalError()
}

let _: AnyHashable = test(overloaded())
```

The conversion is from something `Hashable` to `AnyHashable` so the argument is going to check for `Hashable`

hamishknight replied:
Right, but if it's a different protocol constraint that `AnyHashable` conforms to, but not the `Hashable` type, then I believe that might be problematic.

hborla replied:
Yeah, it's the same issue as `Optional`, `Unsafe*Pointer`, etc

xedin replied:
Ah right, - argument has to be `Hashable` to be implicitly convertible to `AnyHashable` but generic parameter could also require conformance to another type which only `AnyHashable` conforms to...

hamishknight replied:
Looks great! Would it be worth also adding a test case for the conditional conformance example? Admittedly contrived, but another one I thought of was for multiple levels of value-to-optional, e.g:

```
protocol P {}
extension Optional : P where Wrapped == Int? {}

func bar<T : P>(_ x: T) -> T { x }
let i = 0
let _: Int?? = bar(i)
```

xedin replied:
@hamishknight  Sure I can add that! I'm also going to add overloaded version of it to make sure that we handle cases - when simplification is done during constraint generation and when it happens in the solver.

hamishknight reviewed (commented):
Awesome!

xedin commented:
Only one unrelated failure in Kickstarter:

```
rm: /Users/buildnode/jenkins/workspace/swift-PR-source-compat-suite-debug/swift-source-compat-suite/project_cache/Kickstarter-ReactiveExtensions: Directory not empty
```
/cc @shahmishal

shahmishal commented:
> Only one unrelated failure in Kickstarter:
> 
> ```
> rm: /Users/buildnode/jenkins/workspace/swift-PR-source-compat-suite-debug/swift-source-compat-suite/project_cache/Kickstarter-ReactiveExtensions: Directory not empty
> ```
> 
> /cc @shahmishal

@clackary Can you look into this?

xedin committed:
[CSSimplify] Avoid checking conditional conformances in `simplifyTransitivelyConformsTo`

It's not always clear what generic parameter type implicit conversion to pointer
is going to have, and its safer to check direct conformance in `simplifyTransitivelyConformsTo`
and let parameter handle the rest.

xedin committed:
[CSSimplify] Add `AnyHashable` check to `simplifyTransitivelyConformsTo`

Just like `Optional` and `UnsafePointer` argument could be implicitly
converted to `AnyHashable`.

xedin committed:
[CSSimplify] Check if String/AnyHashable/Unsafe{Mutable}Pointer decls are available before using them

xedin committed:
[CSSimplify] Fix transitive conformance check to use `Optional<T>` on if resolved type is not optional

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please smoke test compiler performance

xedin committed:
[TypeChecker] NFC: Add a test-case to test condition requirements interaction with transitive conformance checks

xedin committed:
[TypeChecker] NFC: Add more chained operators to make rdar://22877285 "too complex" on Linux

xedin commented on test/Constraints/protocols.swift:
```diff
@@ -441,11 +441,36 @@ extension UnsafePointer : Trivial {
+  let _: Int?? = simple(i)
+  let _: Int?? = overloaded(i)
+  let _: Int?? = simple(overloaded_result())
+  let _: Int?? = overloaded(overloaded_result())
```
@hborla @hamishknight Added tests for conditional conformances.

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please smoke test compiler performance

xedin commented:
Besides `AnyHashable` we have also forgot about `Unsafe{Mutable}RawPointer` which other pointers and array can convert to, I'll add that soon.

xedin committed:
[CSSimplify] Teach transitive conformance check about Unsafe{Mutable}RawPointer conversions

xedin committed:
[TypeChecker] NFC: Re-enable and move to "fast" tests for rdar://26564101 and rdar://33688063

xedin committed:
[TypeChecker] NFC: Make test-case for rdar://74035425 more complex to avoid flaky behavior

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test Windows platform

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test Windows platform

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test Windows platform
